### PR TITLE
Sureadhere messaging provider integrated

### PR DESCRIPTION
**Pull request by cws1121**

### CODE REVIEW

**Comment by bderenzi** [comment]
@cws1121 - can you run `ruff` on the code to sort the code-style at some point? (fine to wait for further review comments)

File: apps/channels/models.py
Comment (snopoke) [inline apps/channels/models.py:171]: Why use the `client_id` here instead of the `channel.external_id`?

File: apps/channels/tasks.py
Comment (snopoke) [inline apps/channels/tasks.py:66]: Surely it's not possible to create a channel with this platform and a different messaging provider type?

File: apps/channels/tasks.py
Comment (snopoke) [inline apps/channels/tasks.py:69]: I think we should at least log this at the `info` level.

File: apps/channels/tests/message_examples/sureadhere_messages.py
Comment (snopoke) [inline apps/channels/tests/message_examples/sureadhere_messages.py:6]: Just noting that it's a bit odd that the inbound / outbound fields are different:
```diff
- message_text
+ message_Body
- patient_id
+ patient_Id
```

File: apps/chat/channels.py
Comment (snopoke) [inline apps/chat/channels.py:605]: It seems like voice isn't supported so you could leave this property out since the default is `False`

File: apps/service_providers/messaging_service.py
Comment (snopoke) [inline apps/service_providers/messaging_service.py:161]: Is there a way we could reuse these? How long do they last for?

File: apps/service_providers/messaging_service.py
Comment (snopoke) [inline apps/service_providers/messaging_service.py:146]: Is this different for different integrations? Where would one get this from?

File: apps/channels/forms.py
Comment (snopoke) [inline apps/channels/forms.py:75]: Is this `client_id` different from the one on the `MessagingProvider`. If they are the same the it doesn't seem useful to have it here as well.

File: apps/channels/views.py
Comment (snopoke) [inline apps/channels/views.py:35]: limit to POST requests:

```suggestion
@csrf_exempt
@require_POST
```

File: apps/channels/forms.py
Comment (snopoke) [reply apps/channels/forms.py:78]: Can you confirm this understanding:

* `MessagingProvider.client_id`: this is an Oauth client ID used to authenticate with SureAdhere
* `Channel.client_id`: this is a workspace ID in SureAdhere (is a workspace SureAdhere's 'tenent'?)

Is `Client ID` the right name for it then?

It's also not clear to me what this is used for since it isn't included in the requests to send messages so it appears to only be used for the webhook.

I think that leads to the question of how the webhooks in SA are configured. Is there documentation on that somewhere? How does SA know which webhook to call for a specific message?

File: apps/channels/datamodels.py
Comment (snopoke) [inline apps/channels/datamodels.py:108]: Since `content_type` is always `TEXT` I think this can be removed along with `content_type_unparsed`.

File: apps/channels/tests/message_examples/sureadhere_messages.py
Comment (snopoke) [reply apps/channels/tests/message_examples/sureadhere_messages.py:6]: No need to change it. I was just making sure this wasn't a typo or something.

File: apps/channels/views.py
Comment (snopoke) [reply apps/channels/views.py:37]: You still need to add this (you added it to the twilio endpoint but not to this one)

File: apps/service_providers/forms.py
Comment (snopoke) [inline apps/service_providers/forms.py:160]: This should use a URL field:
```suggestion
    base_url = forms.URLField(label=_("Base URL"), validators=[URLValidator(schemes=["https"]))
```

File: apps/channels/forms.py
Comment (snopoke) [reply apps/channels/forms.py]: > The get_success_message method, remains in the respective classes. Let me know if you think we should move it to superclass as well.

I think it should also move since at the moment all the messages are identical. If a class wants to change the message it can override that method.

File: apps/channels/models.py
Comment (snopoke) [reply apps/channels/models.py]: > While SureAdhere is one of the in-app messaging platforms we currently support, what if we decide to add another provider in the future that also offers in-app messaging capabilities?

I'm not sure I quite follow this. If SA adds support for another provider e.g. telegram then OCS will still want to use the SA channel and SA will decide how to deliver that message to the user based on what provider that user is using to communicate with. But OCS won't know about the different downstream providers will it?

File: apps/channels/models.py
Comment (snopoke) [reply apps/channels/models.py]: > However, I agree that using experiment_id

I actually suggested `channel.external_id` since that is the most direct lookup. Is there an advantage to using the experiment ID?

File: apps/channels/models.py
Comment (snopoke) [reply apps/channels/models.py]: I feel like we're missing eachother somewhere and maybe we should go to voice but I'll take one more go at explaining what I think.

When you say "in app", do you mean "communication through a mobile app"? Is that not what WhatsApp and Telegram are already?

In terms of this data model, a channel type should be 1-1 with how the user access it e.g. whatsapp vs telegram vs sureadhere

The `MessagingProvider` ties the channels to a way of sending and receiving messages and is 1-many since the same provider could facilitate communications on multiple channels and multiple providers can support the same channel.

If we take WhatsApp as an example, both Turn and Twilio provide WhatsApp messaging capabilities but which one the user chooses depends on which provider they are using.

If you still have doubts, let's schedule a call.

File: apps/channels/forms.py
Comment (snopoke) [reply apps/channels/forms.py]: Typically webhooks are configured in the sending system so that it know where to send messages. There are 2 options:

1. A generic webhook where the request body contains some information that identifies where the message is coming from e.g. client_id
2. A webhook that contains some identifying info in it e.g. experiment_id or client_id

Either of these approaches are fine but both require some way to configure them in SA.

For now we can revert to using `client_id` since that's what allows you to test but I'd like to hear what the product team has to say.

File: apps/channels/models.py
Comment (snopoke) [inline apps/channels/models.py:95]: this doesn't appear to be used anywhere so you could delete it

File: apps/channels/forms.py
Comment (snopoke) [inline apps/channels/forms.py:78]: Please add some help text here assist the user in understanding what this field is.

**Review by snopoke — COMMENTED** [review]
Thanks for all the updates. This is very close now. I think there are only 2 outstanding things I'd like addressed before approval:

* Revert the webhook to use `client_id` so that you can continue testing
* Add help text to `SureAdhereChannelForm.client_id` indicating what it is

File: apps/channels/forms.py
Comment (SmittieC) [reply apps/channels/forms.py:78]: What's the difference between this `client_id` field and the `client_id` field from the `SureAdhereService`? This client id distinguishes a SureAdhere "project" so to speak whereas the other client id is an Azure id, yes?

File: apps/channels/models.py
Comment (snopoke) [reply apps/channels/models.py:164]: Offlined on this and the webhook calling code in SureAdhere will use a template URL something like `https://ocs-domain.org/channels/sureadhere/{tenant_id}/incoming_message` where the tenant ID will be determined the by the user sending the message.

**Review by snopoke — APPROVED** [review]

**Pull request merged**
